packages/solarwinds-apm/src/storage.ts

```diff
@@ -0,0 +1,163 @@
+ *
+ * @returns Shared global
+ */
+export function global<const T>(id: string, init: () => T): T {
+  const key = Symbol.for(`solarwinds-apm global storage / ${id}`)
+  let storage = Reflect.get(globalThis, key) as T | undefined
+
+  if (!storage) {
+    storage = init()
+    Reflect.defineProperty(globalThis, key, {
+      value: storage,
+      configurable: false,
+      enumerable: false,
+      writable: false,
+    })
+  }
+
+  return storage
+}
+
+/** Typed wrapper around the OTel {@link Context} API for a specific key */
+export interface ContextStorage<T> {
+  /**
+   * Gets the value for the key
+   *
+   * Note that while the context object itself is immutable,
+   * reference types stored in it are not. This means it is possible
+   * to modify the return value of this function in-place if it is
+   * a reference type.
+   */
+  get(context: Context): T | undefined
+  /**
+   * Sets the value for the key
+   *
+   * Note that the context object is immutable and this function
+   * returns an updated context without modifying the given one in-place.
+   * in-place
+   */
+  set(context: Context, value: T): Context
+  /**
+   * Deletes the value for the key
+   *
+   * Note that the context object is immutable and this function
+   * returns an updated context without modifying the given one in-place.
+   */
+  delete(context: Context): Context
+}
+
+/**
+ * Typed OTel {@link Span} attached storage API for a specific key
+ *
+ * This makes it possible to attach arbitrary data to OTel spans
+ * without altering the span object.
+ *
+ * The stored value will be accessible throughout the entire
+ * lifecycle of the span from the moment it is set until the span
+ * finishes being exported.
+ *
+ * Unlike the previous implementation which used the span ID as a key
+ * and needed to be carefully cleared at the right times to avoid leaking
+ * memory, this uses a WeakMap with the span object itself as a key
+ * which ensures the stored values are kept at least as long as they need
+ * to be accessed.
+ *
+ * However it it still useful to manually clear unused entries to avoid
+ * using up memory unnecessarily while waiting for the next GC cycle.
+ */
+export interface SpanStorage<T> {
+  get(span: Span | sdk.Span | sdk.ReadableSpan): T | undefined
+  set(span: Span | sdk.Span | sdk.ReadableSpan, value: T): boolean
+  delete(span: Span | sdk.Span | sdk.ReadableSpan): boolean
+}
+
+const GLOBAL_SPAN_STORAGE = global(
+  "solarwinds-apm span storage",
+  () => new WeakMap<sdk.Span, Map<symbol, unknown>>(),
+)
+
+function withSdkSpan<T, const U>(
+  span: Span | sdk.Span | sdk.ReadableSpan,
+  fallback: U,
+  f: (span: sdk.Span) => T,
+): T | U {
+  if (span instanceof sdk.Span) {
+    return f(span)
+  } else {
+    diag.debug("span storage passed an invalid key", span)
+    return fallback
+  }
+}
+
+/** Creates a new {@link ContextStorage} for the given ID */
+export function contextStorage<T>(id: string): ContextStorage<T> {
```

Comment:
How does contextStorage work? What is the `id` param -- is it the span ID?

Reply:
Ok I've seen now that this is used by the new Headers propagators. Is contextStorage then a wrapper around Otel Context so we can store extras?

Reply:
Yeah that one is just a wrapper to make the Context API accept and return `T` instead of `unknown` :)